DSharpPlus/Net/Rest/DiscordApiClient.cs

```diff
@@ -51,46 +51,50 @@ private DiscordMessage PrepareMessage(JToken msgRaw)
     {
         TransportUser author = msgRaw["author"]!.ToDiscordObject<TransportUser>();
         DiscordMessage message = msgRaw.ToDiscordObject<DiscordMessage>();
-
         message.Discord = this.discord!;
-
         PopulateMessage(author, message);
 
         JToken? referencedMsg = msgRaw["referenced_message"];
-
-        if (message.MessageType == DiscordMessageType.Reply && !string.IsNullOrWhiteSpace(referencedMsg?.ToString()))
+        if (message.MessageType == DiscordMessageType.Reply && referencedMsg is not null && message.ReferencedMessage is not null)
         {
             TransportUser referencedAuthor = referencedMsg["author"]!.ToDiscordObject<TransportUser>();
             message.ReferencedMessage.Discord = this.discord!;
             PopulateMessage(referencedAuthor, message.ReferencedMessage);
         }
 
-        if (message.Channel is not null)
+        return message;
+    }
+
+    private void PopulateMessage(TransportUser author, DiscordMessage ret)
+    {
+        if (ret.Channel is null && ret.Discord is DiscordClient client)
+        {
+            ret.Channel = client.InternalGetCachedChannel(ret.ChannelId);
+        }
+
+        if (ret.guildId is null || !ret.Discord.Guilds.TryGetValue(ret.guildId.Value, out DiscordGuild? guild))
         {
-            return message;
+            guild = ret.Channel?.Guild;
         }
 
-        message.Channel = !message.guildId.HasValue
+        ret.guildId ??= guild?.Id;
+
+        // I can't think of a case where guildId will never be not null since the guildId is a gateway exclusive
+        // property, however if that property is added later to the rest api response, this case would be hit.
+        ret.Channel ??= ret.guildId is null
             ? new DiscordDmChannel
             {
-                Id = message.ChannelId,
+                Id = ret.ChannelId,
                 Discord = this.discord!,
                 Type = DiscordChannelType.Private
             }
             : new DiscordChannel
             {
-                Id = message.ChannelId,
-                GuildId = message.guildId,
+                Id = ret.ChannelId,
+                GuildId = ret.guildId,
                 Discord = this.discord!
             };

```

Comment:
I'm half confident this could be simplified but nothing came to mind after staring at it for 10 minutes